File: app/src/main/java/im/tox/antox/activities/LoginActivity.java
package im.tox.antox.activities;

import android.content.Context;
import android.content.Intent;
import android.content.SharedPreferences;
import android.os.Build;
import android.os.Bundle;
import android.preference.PreferenceManager;
import android.support.v7.app.ActionBarActivity;
import android.view.View;
import android.view.WindowManager;
import android.widget.AdapterView;
import android.widget.ArrayAdapter;
import android.widget.EditText;
import android.widget.Spinner;
import android.widget.Toast;

import java.util.ArrayList;
import java.util.Arrays;

import im.tox.antox.R;
import im.tox.antox.data.UserDB;
import im.tox.antox.tox.ToxDoService;

public class LoginActivity extends ActionBarActivity implements AdapterView.OnItemSelectedListener {

    private String profileSelected;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_login);

        getSupportActionBar().hide();

        /* Fix for an android 4.1.x bug */
        if(Build.VERSION.SDK_INT != Build.VERSION_CODES.JELLY_BEAN
                && Build.VERSION.SDK_INT >= Build.VERSION_CODES.HONEYCOMB) {
            getWindow().setFlags(
                    WindowManager.LayoutParams.FLAG_HARDWARE_ACCELERATED,
                    WindowManager.LayoutParams.FLAG_HARDWARE_ACCELERATED
            );
        }

        SharedPreferences preferences = PreferenceManager.getDefaultSharedPreferences(this);

        UserDB db = new UserDB(this);

        // Check to see if any users exist. If not, start the create account activity instead
        if(!db.doUsersExist()) {
            db.close();
            Intent createAccount = new Intent(getApplicationContext(), CreateAcccountActivity.class);
            startActivity(createAccount);
            finish();

        } else if(preferences.getBoolean("loggedin", false)) {
            db.close();
            /* Attempt to start service in case it's not running */
            Intent startTox = new Intent(getApplicationContext(), ToxDoService.class);
            getApplicationContext().startService(startTox);

            /* Launch main activity */
            Intent main = new Intent(getApplicationContext(), MainActivity.class);
            startActivity(main);

            finish();
        } else {
            ArrayList<String> profiles = db.getAllProfiles();
            db.close();
            // Populate the profile login spinner
            Spinner profileSpinner = (Spinner) findViewById(R.id.login_account_name);
            ArrayAdapter<String> adapter = new ArrayAdapter<String>(
                    this,
                    android.R.layout.simple_spinner_dropdown_item,
                    profiles);
            adapter.setDropDownViewResource(android.R.layout.simple_spinner_dropdown_item);
            profileSpinner.setAdapter(adapter);
            profileSpinner.setSelection(0);
            profileSpinner.setOnItemSelectedListener(this);
        }
    }

    public void onItemSelected(AdapterView<?> parent, View view, int pos, long id) {
        profileSelected = parent.getItemAtPosition(pos).toString();
    }

    public void onNothingSelected(AdapterView<?> parent) {
    }


    public void onClickLogin(View view) {
        String account = profileSelected;

        if (account.equals("")) {
            Context context = getApplicationContext();
            CharSequence text = getString(R.string.login_must_fill_in);
            int duration = Toast.LENGTH_SHORT;
            Toast toast = Toast.makeText(context, text, duration);
            toast.show();
        } else {
            /* Attempt to login */
            UserDB db = new UserDB(this);

            if(db.login(account)) {
                /* Set that we're logged in and active user's details */
                String[] details = db.getUserDetails(account);
                db.close();
                SharedPreferences preferences = PreferenceManager.getDefaultSharedPreferences(this);
                SharedPreferences.Editor editor = preferences.edit();
                editor.putBoolean("loggedin", true);
                editor.putString("active_account", account);
                editor.putString("nickname", details[0]);
                editor.putString("status", details[1]);
                editor.putString("status_message", details[2]);
                editor.apply();

                /* Init Tox and start service */
                Intent startTox = new Intent(getApplicationContext(), ToxDoService.class);
                getApplicationContext().startService(startTox);

                /* Launch main activity */
                Intent main = new Intent(getApplicationContext(), MainActivity.class);
                startActivity(main);

                finish();
            } else {
                Context context = getApplicationContext();
                CharSequence text = getString(R.string.login_bad_login);
                int duration = Toast.LENGTH_SHORT;
                Toast toast = Toast.makeText(context, text, duration);
                toast.show();
            }
        }
    }

    public void onClickCreateAccount(View view) {
        Intent createAccount = new Intent(getApplicationContext(), CreateAcccountActivity.class);
        startActivityForResult(createAccount, 1);
    }

    @Override
    public void onActivityResult(int requestCode, int resultCode, Intent data) {
        if(requestCode == 1) {
            if(resultCode == RESULT_OK) {
                finish();
            }
        }
    }
}

File: app/src/main/scala/im/tox/antox/activities/LoginActivity.scala
package im.tox.antox.activities

import android.app.Activity
import android.content.Intent
import android.graphics.Color
import android.os.{Build, Bundle}
import android.preference.PreferenceManager
import android.support.v7.app.AppCompatActivity
import android.view.{View, WindowManager}
import android.widget._
import im.tox.antox.data.UserDB
import im.tox.antox.tox.ToxService
import im.tox.antoxnightly.R

class LoginActivity extends AppCompatActivity with AdapterView.OnItemSelectedListener {

  private var profileSelected: String = _

  protected override def onCreate(savedInstanceState: Bundle) {
    super.onCreate(savedInstanceState)
    setContentView(R.layout.activity_login)
    getSupportActionBar.hide()
    if (Build.VERSION.SDK_INT != Build.VERSION_CODES.JELLY_BEAN &&
      Build.VERSION.SDK_INT >= Build.VERSION_CODES.HONEYCOMB) {
      getWindow.setFlags(WindowManager.LayoutParams.FLAG_HARDWARE_ACCELERATED, WindowManager.LayoutParams.FLAG_HARDWARE_ACCELERATED)
    }
    val preferences = PreferenceManager.getDefaultSharedPreferences(this)
    val db = new UserDB(this)
    if (!db.doUsersExist()) {
      db.close()
      val createAccount = new Intent(getApplicationContext, classOf[CreateAccountActivity])
      startActivity(createAccount)
      finish()
    } else if (preferences.getBoolean("loggedin", false)) {
      db.close()
      val startTox = new Intent(getApplicationContext, classOf[ToxService])
      getApplicationContext.startService(startTox)
      val main = new Intent(getApplicationContext, classOf[MainActivity])
      startActivity(main)
      finish()
    } else {
      val profiles = db.getAllProfiles
      db.close()
      val profileSpinner = findViewById(R.id.login_account_name).asInstanceOf[Spinner]
      val adapter = new ArrayAdapter[String](this, android.R.layout.simple_spinner_dropdown_item, profiles)
      adapter.setDropDownViewResource(android.R.layout.simple_spinner_dropdown_item)
      profileSpinner.setAdapter(adapter)
      profileSpinner.setSelection(0)
      profileSpinner.setOnItemSelectedListener(this)
    }
  }

  def onItemSelected(parent: AdapterView[_], view: View, pos: Int, id: Long) {
    profileSelected = parent.getItemAtPosition(pos).toString

    if (parent.getChildAt(0) != null) // getChildAt(pos) returns a view, or null if non-existant
      parent.getChildAt(0).asInstanceOf[TextView].setTextColor(Color.BLACK)
  }

  def onNothingSelected(parent: AdapterView[_]) {
  }

  def onClickLogin(view: View) {
    val account = profileSelected
    if (account == "") {
      val context = getApplicationContext
      val text = getString(R.string.login_must_fill_in)
      val duration = Toast.LENGTH_SHORT
      val toast = Toast.makeText(context, text, duration)
      toast.show()
    } else {
      val db = new UserDB(this)
      if (db.doesUserExist(account)) {
        val details = db.getUserDetails(account)
        db.close()
        val preferences = PreferenceManager.getDefaultSharedPreferences(this)
        val editor = preferences.edit()
        editor.putBoolean("loggedin", true)
        editor.putString("active_account", account)
        editor.putString("nickname", details.nickname)
        editor.putString("password", details.password)
        editor.putString("status", details.status)
        editor.putString("status_message", details.statusMessage)
        editor.putBoolean("logging_enabled", details.loggingEnabled)
        editor.putString("avatar", details.avatarName)
        editor.apply()
        val startTox = new Intent(getApplicationContext, classOf[ToxService])
        getApplicationContext.startService(startTox)
        val main = new Intent(getApplicationContext, classOf[MainActivity])
        startActivity(main)
        finish()
      } else {
        val context = getApplicationContext
        val text = getString(R.string.login_bad_login)
        val duration = Toast.LENGTH_SHORT
        val toast = Toast.makeText(context, text, duration)
        toast.show()
      }
    }
  }

  def onClickCreateAccount(view: View) {
    val createAccount = new Intent(getApplicationContext, classOf[CreateAccountActivity])
    startActivityForResult(createAccount, 1)
    overridePendingTransition(android.R.anim.fade_in, android.R.anim.fade_out)
  }

  override def onActivityResult(requestCode: Int, resultCode: Int, data: Intent) {
    if (requestCode == 1) {
      if (resultCode == Activity.RESULT_OK) {
        finish()
      }
    }
  }
}
